### PR TITLE
Enhance logic to deploy server

### LINKED ISSUE #44: Enhance logic to deploy server
* Compute checksum of server binary and use it to determine if deployment is obsolete
* When VSCE starts server, it should detect if binary is missing/outdated and re-deploy
* Optimize deployment to upload only the files that are new or updated

### PR DESCRIPTION
**Pull request by t1m0thyj**
**What It Does**

Resolves #44 
* Compute checksum of server binary and use it to determine if deployment is obsolete
* When VSCE starts server, it should detect if binary is missing/outdated and re-deploy
* Adds unit tests for `ZSshClient` that use Node.js built-in test runner (requires Node 22.x)

For more details about the Node.js test runner, see [this guide](https://betterstack.com/community/guides/testing/nodejs-test-runner/) and docs for the [assert](https://nodejs.org/docs/latest-v22.x/api/assert.html) and [test](https://nodejs.org/docs/latest-v22.x/api/test.html) modules.

**How to Test**

1. Create SSH profile for LPAR with server path that does not exist
2. Add profile to Zowe Explorer tree _without_ "Connect to Host" command
3. Use SSH profile in Zowe Explorer and notice it auto-deploys server
4. Modify one of the checksums in the deployed `checksums.asc` file
5. Reload VS Code, use SSH profile again, and notice it auto-deploys

**Review Checklist**
I certify that I have:
- [ ] tested my changes
- [ ] added/updated automated tests
- [ ] updated the changelog
- [ ] followed the [contribution guidelines](https://github.com/zowe/zowe-cli/blob/master/CONTRIBUTING.md)

**Additional Comments**

### CODE REVIEW

**Comment by zFernand0** [comment]
For those reviewing, you might want to omit the final commit:
- https://github.com/zowe/zowe-native-proto/pull/236/files/3b319069a9b5683c3197686b30b926ddb425927d
 
 This should help 😋

File: scripts/buildTools.ts [hunk 1]
```diff
@@ -290,26 +290,30 @@ async function artifacts(connection: Client) {
...
     for (const localDir of localDirs) {
         mkdirSync(resolve(__dirname, `./../${localDir}`), { recursive: true });
-        if (localDirs.indexOf(localDir) === 0) {
-            await retrieve(connection, [paxFile], localDir);
-        } else {
-            cpSync(
-                resolve(__dirname, `./../${localDirs[0]}/${paxFile}`),
-                resolve(__dirname, `./../${localDir}/${paxFile}`),
-            );
+        for (const localFile of [paxFile, checksumFile]) {
```
Comment (zFernand0) [inline scripts/buildTools.ts:308]: What if we need to deploy a few more files in the future? (e.g. a separate pax file that contains only the APF authorized version - `zoweax`)

In order to minimize the number of places we need to look for when adding new files... could we create this array towards the top of the function? 😋

File: packages/vsce/src/SshClientCache.ts [hunk 2]
```diff
@@ -28,8 +32,12 @@ export class SshClientCache extends vscode.Disposable {
         }
     }
 
+    public static initialize(context: vscode.ExtensionContext): SshClientCache {
+        SshClientCache.mInstance = new SshClientCache(context);
```
Comment (zFernand0) [inline packages/vsce/src/SshClientCache.ts:36]: I can see how it might be easier to manipulate (and test) the instance when it's initialized via a function instead of a getter. I do have one question... What about keeping the "singleton" approach?

```suggestion
        SshClientCache.mInstance ??= new SshClientCache(context);
```

And perhaps making this function `@internal` 😋

File: packages/vsce/src/SshClientCache.ts [hunk 3]
```diff
@@ -38,22 +46,43 @@ export class SshClientCache extends vscode.Disposable {
...
+                    onConnect: async (client: Client) => {
+                        if (
+                            autoUpdate &&
+                            (await ZSshUtils.checkIfOutdated(client, serverPath, path.join(localDir, "checksums.asc")))
+                        ) {
+                            imperative.Logger.getAppLogger().info(
+                                `Server is out of date, deploying to ${profile.name}`,
+                            );
+                            await deployWithProgress(session, serverPath, localDir);
+                        }
```
Comment (zFernand0) [inline packages/vsce/src/SshClientCache.ts:71]: could we have an `else` to at least warn the user about the server being out of date (in case they turn off the setting) 😋

**Review by zFernand0 — COMMENTED** [review]
Changes LGTM! 😋 

Still reviewing though 😋

File: scripts/buildTools.ts
[the remark this answers is not in the document]
Comment (t1m0thyj) [reply scripts/buildTools.ts:1]: Resolved in d2f33e78908858575e491a1fe0c9d307ef478218

File: scripts/buildTools.ts (hunk 1 above)
Comment (t1m0thyj) [reply scripts/buildTools.ts]: Resolved in d2f33e78908858575e491a1fe0c9d307ef478218

File: packages/vsce/src/SshClientCache.ts (hunk 2 above)
Comment (t1m0thyj) [reply packages/vsce/src/SshClientCache.ts:36]: Discussed in person, I believe this method is fine to leave as-is since we need to pass `context`, and we don't need to mark it internal since this class is not exported.

File: biome.json
Comment (traeok) [inline biome.json:1]: Are the tabs intentional? 😋 Just curious as I've been using spaces for indentation, but wasn't sure what the team preferred

File: native/golang/main.go [hunk 4]
```diff
@@ -75,9 +75,16 @@ func main() {
...
+	var data *map[string]any = nil
+	if checksums := utils.LoadChecksums(); checksums != nil {
+		data = &map[string]any{
+			"checksums": checksums,
+		}
+	}
 	readyMsg := t.StatusMessage{
 		Status:  "ready",
 		Message: "zowed is ready to accept input",
+		Data:    data,
```
Comment (traeok) [inline native/golang/main.go:87]: Can we avoid the use of a pointer here?

```suggestion
	data := make(map[string]any)
	if checksums := utils.LoadChecksums(); checksums != nil {
		data["checksums"] = checksums
	}
	readyMsg := t.StatusMessage{
		Status:  "ready",
		Message: "zowed is ready to accept input",
		Data:    data,
```

I guess the difference here is that data would be an empty object on the response with this approach (`{}` vs. `undefined`), so maybe that isn't the preferred behavior 😋 

Alternatively, we could assign the value of `utils.LoadChecksums` directly to a key in the map, and when the checksums are absent we know based on the serialization of `nil`:

```suggestion
	data := make(map[string]any)
	data["checksums"] = utils.LoadChecksums()
	readyMsg := t.StatusMessage{
		Status:  "ready",
		Message: "zowed is ready to accept input",
		Data:    data,
```

File: native/golang/utils/bytes.go [hunk 5]
```diff
@@ -44,3 +46,24 @@ func CollectContentsAsBytes(input string, isByteString bool) (data []byte, err e
...
 	return
 }
+
+func LoadChecksums() map[string]string {
+	checksumsFile := "checksums.asc"
+	file, err := os.Open(checksumsFile)
+	if os.IsNotExist(err) {
+		return nil
+	} else if err != nil {
+		panic(err)
```
Comment (traeok) [inline native/golang/utils/bytes.go:56]: Can we add a `utils.LogError` here to track weird edge-cases, specifically there is an error that isn't related to the absence of a checksums file?

**Review by traeok — CHANGES_REQUESTED** [review]
Thanks for adding tests! This looks great 😋 working on testing now, but I have a small request for logging and a question about the use of a pointer type in `StatusMessage`

File: biome.json
Comment (t1m0thyj) [reply biome.json:1]: Biome formats only our TypeScript code. I've also noticed the Golang code having tabs instead of spaces, which is different from what I'm used to but I think that's coming from the formatter of the Go VSCE.

File: native/golang/main.go (hunk 4 above)
Comment (t1m0thyj) [reply native/golang/main.go]: Good suggestion, addressed in d69370048d4e26e3761bebee3cdfe1fddb7cd2c1

**Review by traeok — APPROVED** [review]
LGTM, thanks Timothy for the enhancements!
